### ISSUE TITLE
uploadFile error 4.0.15

### ISSUE THREAD

**Issue by timershu** [netty/netty#3303]
Some file upload will be in error, some not
![netty-file](https://cloud.githubusercontent.com/assets/8155279/5609977/67d41d5c-94e8-11e4-982a-5ad697c82323.png)

public static boolean uploadFile(Bootstrap b,final File file,String host,int port) throws Exception {
    	HttpDataFactory factory = new DefaultHttpDataFactory(DefaultHttpDataFactory.MINSIZE);
        if (!file.canRead()) {
            throw new FileNotFoundException(file.getAbsolutePath()+"/"+file.getName());
        }
        URI uriFile = new URI("/formpost/zhenai" );
        // Start the connection attempt.
        final ChannelFuture cf = b.connect(new InetSocketAddress(host, port));
        // Connection established successfully
        // Wait until the connection attempt succeeds or fails.
        Channel channel = cf.sync().channel();

        System.out.println("uriFile.toASCIIString()=="+uriFile.toASCIIString());
        // Prepare the HTTP request.
        HttpRequest request = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.POST, uriFile.toASCIIString());

        // Use the PostBody encoder
        HttpPostRequestEncoder bodyRequestEncoder =
                new HttpPostRequestEncoder(factory, request, true); // true => multipart

      
        // add Form attribute
        bodyRequestEncoder.addBodyFileUpload("file", file, "application/x-zip-compressed", false);
        // finalize request
        try {
            request = bodyRequestEncoder.finalizeRequest();
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        // send request
        channel.write(request);
        // test if request was chunked and if so, finish the write
        if (bodyRequestEncoder.isChunked()) {
        	ChannelFuture cFuture=channel.writeAndFlush(bodyRequestEncoder);
        } else {
            channel.flush();
        }
        // Now no more use of file representation (and list of HttpData)
        bodyRequestEncoder.cleanFiles();
        // Wait for the server to close the connection.
        channel.closeFuture().sync();
        //callback.call(true);
        return true;
    }

**Comment by normanmaurer** [comment]
@timershu please upgrade to 4.0.25.Final and try again. I think this was fixed before. If this is still an issue re-open.

**Comment by timershu** [comment]
There is an exception, some file upload, but the contents of the documents not upload to the server

**Comment by timershu** [comment]
![netty01](https://cloud.githubusercontent.com/assets/8155279/5610154/bc00886c-94ec-11e4-8b41-7cb9096179e8.png)

**Comment by timershu** [comment]
@normanmaurer  Change package, bug still exists, the abnormal information is as follows。
![netty02](https://cloud.githubusercontent.com/assets/8155279/5639270/1ff9b174-964e-11e4-8d68-9a1af21624ab.png)

**Comment by normanmaurer** [comment]
@timershu we will have a look.... /cc @trustin

**Comment by normanmaurer** [comment]
@timershu I think it's because you call cleanFiles() before the actual response may be written.

Use this:
```java
// test if request was chunked and if so, finish the write
        if (bodyRequestEncoder.isChunked()) {
            ChannelFuture cFuture=channel.writeAndFlush(bodyRequestEncoder);
            cFuture.addListener(new ChannelFutureListener() {
                @Override
                public void operationComplete(ChannelFuture future) throws Exception {
                    // Now no more use of file representation (and list of HttpData)
                    bodyRequestEncoder.cleanFiles();
                }
            });
        } else {
            channel.flush();
            bodyRequestEncoder.cleanFiles();
        }
```

**Comment by timershu** [comment]
@normanmaurer Because not too stable, we decided not to use this scheme.

**Comment by normanmaurer** [comment]
@timershu I don't understand... Can you explain what you try to say ?

**Comment by timershu** [comment]
Because not too stable, we decided not to use this program.

**Comment by timershu** [comment]
@normanmaurer  In the day with 1000 files to be uploaded, occasionally only individual file upload failed

**Comment by normanmaurer** [comment]
@timershu I still not get your point... the exception you posted here is because you call cleanFiles() to early . That will be fixed by my change.

**Comment by timershu** [comment]
@normanmaurer  I would like to participate in this project, but the leaders feel that is not too stable, you need a change of plan. This random bug is how to produce? To provide under the relevant design data?

**Comment by timershu** [comment]
@normanmaurer  I am in accordance with you provided inside the demo to write

**Comment by timershu** [comment]
@normanmaurer  
![netty03](https://cloud.githubusercontent.com/assets/8155279/5656377/1ff6773e-9716-11e4-8a1e-ccb02464f188.png)
![netty04](https://cloud.githubusercontent.com/assets/8155279/5656378/24059332-9716-11e4-9142-e360fa026ba0.png)

**Comment by timershu** [comment]
@normanmaurer  @daschl This problem is whether the individual files, file reason?

**Comment by timershu** [comment]
![netty05](https://cloud.githubusercontent.com/assets/8155279/5659773/0e6d67f6-9753-11e4-9235-cc2a6c87d5c0.png)

**Comment by cleverpig** [comment]
mark~we have fixed in 4.0.26?
